### DISCUSSION TITLE
How can I view multiple versions side by side?

### DISCUSSION THREAD

**Discussion by joyously** [jj-vcs/jj#3515] [Q&A]
I have a small Git repo that I was working on last year. Apparently, I was lax in committing my changes and ended up with a couple of copies that I need to sort out. 

I'm looking for a `jj` command that can feed consecutive commits to an external diff tool. I have Diffuse which can handle n-way diffs, but also Meld which says it can do 3. Maybe there is a better way/tool. 

What I'm after is a detailed view of how the code evolved. This use-case is for forgotten projects as well as coming up to speed on new assignments. Oh, it seems like it could also be good for reviews in some workflows. (Maybe this is what `blame` could be?)

Do I need a script to extract a bunch of versions and then load those into Diffuse?
I tried using Meld with `jj log -p` and it's cool, but it only shows one change at a time and no labels for which change you are looking at (inside Meld).

By the way, the docs are sparse for the commands. I did see `diff` and it has `--from` and `--to`, but I don't think it means all the ones between also, which is what I would want, since the first line says 
> Compare file contents between two revisions

Then there is `interdiff` with a very similar first line
> Compare the changes of two commits

but it does some sort of rebase.
> temporarily rebasing `--from` onto `--to`'s parents.

It also says something tantalizingly close (I think), but it makes no sense to me.
> If you wish to compare the same change across versions, consider `jj obslog -p` instead.

I'm going to guess that starting with `jj` on an old repo won't have anything in `obslog`.
And then there is `log`, which seems promising, if I could just get to see more than one at a time.

**Comment by ilyagr** [comment]
AFAIK, there is no easy way to do that now. In principle, it would be quite feasible to implement a command to launch `diffuse` in this way, but:

- `diffuse` is (unfortunately) the only graphical tool I'm aware of that supports comparing multiple files. (vimdiff could do it too, in principle, but it wouldn't be nice to use).
- Diffuse is not very popular, not easy to install on Mac OS, and implementing this is not trivial (one has to think about how the config and the UI should look).

So, I'm not planning on implementing something like this anytime soon. It'd be a fun feature if somebody did implement it.

--------------

If you seriously need this, one thing you could do is to use `jj workspace add -r <REVISION> dir` for each revision to create a separate working copy for each revision you care about. You could then try using Diffuse to compare these directories (assuming it can compare multiple directories, which I hope it can). 

**Update:** It doesn't seem it can compare directories, so that's an additional difficulty.

You could indeed make a script to do this.

Fun fact: this is very similar to what `jj` does automatically when launching an interactive editing session.


--------------

I had some hopes for my [`diffedit3`](https://github.com/ilyagr/diffedit3) to be second such tool after `diffuse`, but the library I'm using doesn't support more than  3-way comparisons. So, it'd require re-implementing the part of the library that does the UI for the comparison. I should mention this in the README, in case there are other reasons to re-implement parts of that library.
